webcompat/static/js/lib/models/query-params.js

```diff
@@ -0,0 +1,360 @@
+            this.set(name, [], {silent:true});
+            for(var i = 0; i < value.length; i++) {
+              this.setParam(name, value[i], silent)
+            }
+          } else { // We add the new value to the existing array
+            this.set(name, currentValue.concat([value]), options);
+          }
+        } else {
+          this.set(name, [], options);
+        }
+      } else {
+        this.set(name, value, options);
+      }
+      return;
+    } else {
+      // if it's not a known property, stuff it into search
+      // append name:value to existing search
+      // TODO: we should never get here. Is it OK to stuff it into search,
+      // or should we throw?
+      this.set('q', this.get('q') + ' ' + name + ':' + value, options);
+    }
+    if(silent) {
+      // The UI is normally updated after a successful request to the backend,
+      // however if something is broken and the request doesn't suceed, certain
+      // parts of the UI that are changed directly by the user will now be out of
+      // sync with the model's state. In particular, if the search field was edited
+      // and the request failed, the model still remembers the search query but it's
+      // not in the UI.
+      // It may be somewhat annoying and surprising that a search term you've changed
+      // reverts to its previous state if the backend acts up - but the alternative is
+      // worse: UI inconsistency, with a list that doesn't match the search UI state.
+      issueList.events.trigger('search:update', this.get('q'));
+    }
+  },
+  deleteLabel: function(labelStr){
+    var currentLabels = this.get('labels');
+    if(currentLabels.indexOf(labelStr) > -1) {
+      currentLabels.splice(currentLabels.indexOf(labelStr), 1);
+      this.set('labels', {}, {silent:true}); // hack: to trigger change event on *next* set..
+      // hack explained: currentLabels is a reference to an array, so updates cause immediate change.
+      // However, when we're manipulating the array directly, the model does not fire any change events
+      // because JS gives Backbone no way to observe changes to an array. If we set label to a reference
+      // to the array it already references, it won't fire change events either, even if array content
+      // was updated in the meantime. Hence the workaround sets it to {} and then back to the array.
+      this.setParam('labels', currentLabels);
+    }
+  },
+  fromQueryString: function(str, silent){
+    var self = this, namevalues, pair, name, value;
+    if(str.substr(0,1) === '?') {
+      str = str.substr(1);
+    }
+    namevalues = str.split(/&/g);
+    namevalues.forEach(function(namevalue){
+      pair = namevalue.split('=');
+      name = pair[0];
+      value = pair[1];
+      // just in case..
+      if(pair[0] === 'label') {
+        pair[0] = 'labels';
+      }
+      self.setParam(pair[0], pair[1]);
+    });
+  },
+  // silentReset is intended used when a request to the backend fails
```

Comment:
This is mainly included because I noticed such bugs very well when at Mozlando :)